tests/Console/ConfigurationResolverTest.php

```diff
@@ -868,6 +869,123 @@ public function testResolveRulesWithOption()
         );
     }
 
+    /**
+     * @param string[] $rules
+     * @param string[] $negativeRules
+     *
+     * @dataProvider provideSubtractionCases
+     */
+    public function testResolveRulesWithNegativeSet(array $rules, array $negativeRules)
```

Comment:
Looking at the code; to do it the nice way we should move the rule set definitions out of the `RuleSet` class and introduce something like `RuleSetDefinitions` or `RuleSetRepository` and inject that into `RuleSet`. Maybe even better to introduce a `RuleSetResolver` and inject the `RuleSet` and `RuleSetDefiniton` and let the resolver figure out what the set of rules should be.
Currently I don't have the time to make a POC on this though, so maybe park this PR for a bit.